### PR TITLE
[4.0] Blog Sampledata: Fix links, use  camelCase for variable catIds

### PR DESCRIPTION
**Pull request by chmst**
Pull Request for Issue #31690 ., #32270

### Summary of Changes

- Fix double id=id=  
- Change url in banner module as required.
- Use camelCaps for a variable catIds. 
- Change the title "Admin Menu" to "Special Menu".
- Change the the "Author Login" to "Administrator Login" in Text.
- Use access levels for changing login / logout links in top menu and footer menu
- Don't enable workflow automatically
- Small code improvements

### Testing Instructions
Make a new installation of j4 and install the patchtester. (thanks @richard67)
To make testing easier: export the database (with drop table) , so you can import it for the second step.

Install Blog Sample Data in a fresh J4 and refresh the dashhboard view
Find all the issues described in "BEFORE Patch" 

Apply the Patch.
Import now the database which you have saved (or make a new installation)
Find all the resolved issues described in "AFTER Patch" 

### Actual result BEFORE applying this Pull Request
Backend: 
- In the menu "content", the menu item workflow is visible (after refresh), the workflow is enabled
- There is a  menuItem anf a module "Author menu"
- There are some menuItems where the link contains "id=id="

Frontend:
- The link in the banner to typography is not made ser friendly
- In the top menu and bottom menu there is a link "login". It remains login, also after you logged in.
- If the user logs out, a 404 occurs

### Expected result AFTER applying this Pull Request
Backend: 
- No menuItem workflow is visible (after refresh)
- Menu Item and module "Author menu" are now "Special menu" 
- In all menuItems  "id=id=" is correct "id="
- Check the menuItems in Admin Menu. There must be a consitent tree structure, no orphaned items 

Frontend:
- The link in the banner to typography is not made userfriendly.
- If a user is logged in, the menu Items logout appear and vice versa.
- If the user logs out, no error occurs

### CODE REVIEW

**Comment by brianteeman** [comment]
The menu items are correct now but the typography link is now index.php?itemid=106

File: plugins/sampledata/blog/blog.php [hunk 1]
```diff
@@ -988,7 +988,7 @@ public function onAjaxSampledataApplyStep2()
 				// Typography
 				'menutype'     => $menuTypes[0],
 				'title'        => Text::_('PLG_SAMPLEDATA_BLOG_SAMPLEDATA_MENUS_ITEM_14_TITLE'),
-				'link'         => 'index.php?option=com_content&view=article&id=&id=' . (int) $articleIds[10],
+				'link'         => 'index.php?option=com_content&view=article&id=' . (int) $articleIds[10] ,
```
Comment (Quy) [inline plugins/sampledata/blog/blog.php]:
```suggestion
				'link'         => 'index.php?option=com_content&view=article&id=' . (int) $articleIds[10],
```

**Comment by brianteeman** [comment]
the $headerLink needs to be something like
`		$headerLink = 'index.php?option=com_content&view=article&id=' . $articleIds[10] . '&catid=' . $catIds[4];
`
but I couldnt get the value of the catid. Must be missing something obvious.

**Comment by chmst** [comment]
@brianteeman changed as desired. The categoryIds were missing in step 3 ...<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31711">issues.joomla.org/tracker/joomla-cms/31711</a>.</sub>

File: plugins/sampledata/blog/blog.php [hunk 2]
```diff
@@ -1368,14 +1373,16 @@ public function onAjaxSampledataApplyStep3()
...
 
+		// Get previously entered Data from UserStates.
+		$articleIds = $this->app->getUserState('sampledata.blog.articles');
+
 		// Get previously entered Data from UserStates
 		$menuTypes = $this->app->getUserState('sampledata.blog.menutypes');
 
-		$catids     = $this->app->getUserState('sampledata.blog.articles.catids');
-		$articleIds = $this->app->getUserState('sampledata.blog.articles');
+		$catIds        = $this->app->getUserState('sampledata.blog.articles.catIds');
```
Comment (richard67) [inline plugins/sampledata/blog/blog.php]:
```suggestion
		$catIds = $this->app->getUserState('sampledata.blog.articles.catIds');
```

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 42362d832b8718e6fa12f9fb006236a2e5374495<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31711">issues.joomla.org/tracker/joomla-cms/31711</a>.</sub>

**Comment by richard67** [comment]
@brianteeman I'd be happy if you could find some time and test @chmst 's PR here. Thanks in advance.

**Comment by infograf768** [comment]
Title missing for Special Menu Module

Screen Shot 2021-02-13 at 18 27 00

**Comment by chmst** [comment]
Use access levels for changing login / logout links in top menu and footer menu #32556

**Comment by richard67** [comment]
@chmst Does your last change close issue #32556 ?

**Comment by chmst** [comment]
@infograf768 this was intended as example for a module without title.

**Comment by chmst** [comment]
Don't activate workflow automatically. #32387

File: plugins/sampledata/blog/blog.php [hunk 3]
```diff
@@ -11,6 +11,7 @@
 
 use Joomla\CMS\Application\ApplicationHelper;
 use Joomla\CMS\Component\ComponentHelper;
+use Joomla\CMS\Plugin\PluginHelper;
 use Joomla\CMS\Extension\ExtensionHelper;
 use Joomla\CMS\Factory;
```
Comment (Quy) [inline plugins/sampledata/blog/blog.php]:
```suggestion
use Joomla\CMS\Extension\ExtensionHelper;
use Joomla\CMS\Factory;
use Joomla\CMS\Plugin\PluginHelper;
```

File: plugins/sampledata/blog/blog.php [hunk 4]
```diff
@@ -13,6 +13,7 @@
 use Joomla\CMS\Component\ComponentHelper;
 use Joomla\CMS\Extension\ExtensionHelper;
 use Joomla\CMS\Factory;
+use Joomla\CMS\Plugin\PluginHelper;
```
Comment (Quy) [inline plugins/sampledata/blog/blog.php]: Sorry, please move it after line 20.

Comment (chmst) [reply plugins/sampledata/blog/blog.php]: np, hopefully it is ok now

**Comment by sksuryan** [comment]
While trying to login, facing a 404.

image


Steps to reproduce:
- Go the landing page.
- Click on login tab.
- Use the login form on the right.

image

***Side notes**: The same issue doesn't seem to happen with the form on the left. The form on the left redirects to home page.*

Details about system:
I'm currently on `Windows 10`, using `WAMP server`. I can confirm this issue for both latest `Mozilla Firefox` and `Microsoft Edge`.

I used a fresh installation of Joomla 4.0.8 Nightly with a new database. Next, I installed the patch tester and applied the patch. Lastly, I installed the Sample Blog Data.

The expected behavior is that the user is redirected to either home page or profile page.

**Comment by chmst** [comment]
@sksuryan thank you for testing. I have repeated your test and it is only if the active view is the login view, right?

**Comment by sksuryan** [comment]
Yes. 
Also, I'm a little confused. I don't understand why do we need two login forms? If I'm not wrong, the form on the right is the part of the standard sidebar which we see on every page, but it kind of looks redundant here.

**Comment by drmenzelit** [comment]
> 
> 
> Yes.
> Also, I'm a little confused. I don't understand why do we need two login forms? If I'm not wrong, the form on the right is the part of the standard sidebar which we see on every page, but it kind of looks redundant here.

You don't need to have the two... the one is a module, the other is the view of a component. You can use whatever suit better for you, but in sample data we have the two alternatives to show what is possible.

**Comment by infograf768** [comment]
> If I'm not wrong, the form on the right is the part of the standard sidebar which we see on every page, but it kind of looks redundant here.

That login module may be published or not and, if published, not on every page.
Also, when one is using, for example, a readmore which needs user to be logged, it is the login menu item form which is presented to the user to fill, and not the module. ;)

**Comment by sksuryan** [comment]
> 
> 
> > Yes.
> > Also, I'm a little confused. I don't understand why do we need two login forms? If I'm not wrong, the form on the right is the part of the standard sidebar which we see on every page, but it kind of looks redundant here.
> 
> You don't need to have the two... the one is a module, the other is the view of a component. You can use whatever suit better for you, but in sample data we have the two alternatives to show what is possible.



> 
> 
> > If I'm not wrong, the form on the right is the part of the standard sidebar which we see on every page, but it kind of looks redundant here.
> 
> That login module may be published or not and, if published, not on every page.
> Also, when one is using, for example, a readmore which needs user to be logged, it is the login menu item form which is presented to the user to fill, and not the module. ;)

okay, yes that actually makes a lot of sense. thank you!

**Comment by chmst** [comment]
It is confusing. 
When Joomla is installed, a main menu (home) and a login module are stored. This is independent of blog sample data. I don't know if there is a reason in history why the login module is added during installation. As @infograf768 says, login and logout is  necessary, but if it is needed for installaition?

Blog Sample Data are optional and should show how modules can be added to front-and backend, and especially how Login and Logout can be changed using user groups.

**Comment by brianteeman** [comment]
> I don't know if there is a reason in history why the login module is added during installation.

No good reason

**Comment by chmst** [comment]
To resolve the https://github.com/joomla/joomla-cms/pull/31711#issuecomment-798335650 the login module is now hidden on login view.

**Comment by srishty-07** [comment]
I have tested this item :white_check_mark: successfully on 88293c3cd4831e8e0a1d50f2a5851adb71acaaec<br /><br />Tested it successfully on Joomla! 4.0.0-beta8-dev ,php 8.0.0 <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31711">issues.joomla.org/tracker/joomla-cms/31711</a>.</sub>

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 88293c3cd4831e8e0a1d50f2a5851adb71acaaec<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31711">issues.joomla.org/tracker/joomla-cms/31711</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31711">issues.joomla.org/tracker/joomla-cms/31711</a>.</sub>

**Comment by chmst** [comment]
Back to pending - the admin menu looks strange and needs new tests.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31711">issues.joomla.org/tracker/joomla-cms/31711</a>.</sub>

**Comment by ghost** [comment]
@chmst Some Questions about "Actual result BEFORE applying this Pull Request"

### Backend:
> - There is a menuItem anf a module "Author menu"

Found only menu "Admin Menu", module has title "Author menu".

> - There are some menuItems where the link contains "id=id="

Found this menus, are there more?
1. About your home page, id=119
2. Millions, id=121
3. Love, id=122
4. Typography, id=106

### Frontend:
> - If the user logs out, a 404 occurs

How to get a 404, https://github.com/joomla/joomla-cms/pull/31711#issuecomment-798335650 don't bring 404 up. 
Test Login and -out at Menu `Home` and Module `Login Form` didn't get the 404. Maybe solved elsewhere?

#### Sorry for wrong test, i test only issue #32991.

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 4d3ac2304643d0c3a29e1148b8e650c829569e18<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31711">issues.joomla.org/tracker/joomla-cms/31711</a>.</sub>

**Comment by ghost** [comment]
~@chmst can you please answer https://github.com/joomla/joomla-cms/pull/31711#issuecomment-813257534 so i can do a test?~

**Comment by sksuryan** [comment]
I have tested this item :white_check_mark: successfully on 4d3ac2304643d0c3a29e1148b8e650c829569e18<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31711">issues.joomla.org/tracker/joomla-cms/31711</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31711">issues.joomla.org/tracker/joomla-cms/31711</a>.</sub>

**Comment by chmst** [comment]
@sandramay0905 thank you very much for testing and sorry for delay.

>Found only menu "Admin Menu", module has title "Author menu".

Correct. The name of the module must not be the same as the link. The title of the module is not shown in the frontend, this is intended.

>Found this menus, are there more?

There were 4 menuItems - but which IDs they have depends on your installation. A user  can work in an application before adding the sample data and then IDs are different.

>Test Login and -out at Menu Home and Module Login Form didn't get the 404. Maybe solved elsewhere?

I also cannot reproduce this effect with 404. It leads me now to the profile page (which is nearly as bad as 404). So probably one of the recent PRs has fixed this.

**Pull request merged**